### PR TITLE
Added custom loss function that can be defined externally.

### PR DESCRIPTION
**Pull request by prince776**
Hi, I've added Loss layer for which forward and backward functions can be defined externally.
This will allow users to test their own loss functions instead of just being able to use predefined ones.

**Why I did this?**
Keras and Pytorch also have same functionality, except the fact we only need to define forward propagation in them, but in this implementation, user will need to define backward propagation method too.

**How to use this loss layer?**
To implement this, user will have to define his own `double forward(const mat&& input, const mat&& target)` and `void backward(const mat&& input, const mat&& target, mat&& output)` functions.

**Other method I tried**
I also did it in a way such that these functions to be defined only perform calculations on single input and target and not whole matrices, but then that would make user unable to create loss functions like cross entropy for activations like softmax.

Please give any suggestions and review to improve the implementation.

### CODE REVIEW

File: src/mlpack/methods/ann/loss_functions/custom_loss.hpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+   * @param backward custom backward function for the loss that takes in
+   *        input, target and output matrices as parameter and saves the
+   *        gradient in the ouput matrix.
+   */
+  CustomLoss(
+      std::function<double(const InputDataType&&,
+                           const InputDataType&&)> forward,
+      std::function<void(const InputDataType&&,
+                         const InputDataType&&, OutputDataType&&)> backward);
+  /**
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/loss_functions/custom_loss.hpp]: Could you add an empty line here.

**Comment by kartikdutt18** [comment]
Also, I think mlpack has support for custom layers, If you could also provide link for that, to compare why this is needed. It was added as a milestone for mlpack 3.0.0. I will also search for the same in the evening.

**Comment by prince776** [comment]
> Also, I think mlpack has support for custom layers, If you could also provide link for that, to compare why this is needed. It was added as a milestone for mlpack 3.0.0. I will also search for the same in the evening.

Thanks for the review @kartikdutt18 , but I'm unable to find where custom layer support is. 
I didn't find it earlier that's why I implemented it.

**Comment by kartikdutt18** [comment]
Hmm, I'll try finding it because if you refer #1160, It has been implemented.

**Comment by rcurtin** [comment]
I'm not sure this is needed. It's virtually identically easy to simply define the custom layer as a class.  I think this is discussed in the neural network tutorial (not 100% sure).

**Comment by prince776** [comment]
> I'm not sure this is needed. It's virtually identically easy to simply define the custom layer as a class. I think this is discussed in the neural network tutorial (not 100% sure).

Yes the method described in tutorial seems almost identically easy. My bad I didn't look through it and implemented this. So should I close this PR or should it be added as an additional feature since I feel like it would be easier to just define two functions than a whole class, but this is also limited to only loss functions.

**Comment by rcurtin** [comment]
I think that we should close it.  We should avoid adding code that we don't need to (that's just more maintenance load for later).

**Comment by prince776** [comment]
> I think that we should close it. We should avoid adding code that we don't need to (that's just more maintenance load for later).

Ok thanks for your input.
